// thing-it-device-kidde/smokeDetector.js
module.exports = {
    metadata: {
        family: "smokeDetector",
        plugin: "smokeDetector",
        label: "Kidde © Smoke Detector",
        tangible: true,
        discoverable: true,
        state: [{
            id: "batteryLevel",
            label: "Battery Level",
            type: {
                id: "number"
            }
        }, {
            id: "particleConcentration",
            label: "Particle Concentration",
            type: {
                id: "number"
            }
        }],
        events: [{id: "alarm", label: "Alarm"}],
        actorTypes: [],
        sensorTypes: [],
        services: [],
        configuration: [{
            id: "particleConcentrationThreshold",
            label: "Particle Concentration Threshold",
            type: {
                id: "number"
            }
        }]
    },
    create: function () {
        return new SmokeDetector();
    },
    discovery: function (options) {
        var discovery = new SmokeDetectorDiscovery();

        discovery.options = options;

        return discovery;
    }
};

var q = require('q');

function SmokeDetectorDiscovery() {
    /**
     *
     * @param options
     */
    SmokeDetectorDiscovery.prototype.start = function () {
        if (this.node.isSimulated()) {
        } else {
        }
    };

    /**
     *
     * @param options
     */
    SmokeDetectorDiscovery.prototype.stop = function () {
    };
}

/**
 *
 */
function SmokeDetector() {
    /**
     *
     */
    SmokeDetector.prototype.start = function () {
        var deferred = q.defer();

        this.state = {consumption: 0};

        if (this.isSimulated()) {
            setInterval(function () {
                this.state.batteryLevel = 100 - 20 * Math.floor((Math.random() * 6));
                this.state.particleConcentration = 0.4 + 0.1 * Math.floor((Math.random() * 3));

                this.publishStateChange();

                var threshold = this.configuration && this.configuration.particleConcentrationThreshold ? this.configuration.particleConcentrationThreshold : 0.5;

                if (threshold) {
                    this.publishEvent("alarm");
                }
            }.bind(this), 10000);
            deferred.resolve();
        } else {
            deferred.resolve();
        }

        return deferred.promise;
    };

    /**
     *
     */
    SmokeDetector.prototype.setState = function (state) {
        this.state = state;

        this.publishStateChange();
    };

    /**
     *
     */
    SmokeDetector.prototype.getState = function () {
        return this.state;
    };
}
