// thing-it-device-plugwise/switch.js
module.exports = {
    metadata: {
        family: "switch",
        plugin: "switch",
        label: "Plugwise © Smart Switch",
        tangible: true,
        discoverable: true,
        state: [{
            id: "switch",
            label: "Switch",
            type: {
                id: "boolean"
            }
        }],
        actorTypes: [],
        sensorTypes: [],
        services: [{
            id: "on",
            label: "On"
        }, {
            id: "off",
            label: "Off"
        }, {
            id: "toggle",
            label: "Toggle"
        }],
        configuration: []
    },
    create: function () {
        return new Switch();
    },
    discovery: function (options) {
        var discovery = new SwitchDiscovery();

        discovery.options = options;

        return discovery;
    }
};

var q = require('q');

function SwitchDiscovery() {
    /**
     *
     * @param options
     */
    SwitchDiscovery.prototype.start = function () {
        if (this.node.isSimulated()) {
        } else {
        }
    };

    /**
     *
     * @param options
     */
    SwitchDiscovery.prototype.stop = function () {
    };
}

/**
 *
 */
function Switch() {
    /**
     *
     */
    Switch.prototype.start = function () {
        var deferred = q.defer();

        this.state = {switch: false};

        if (this.isSimulated()) {
            deferred.resolve();
        } else {
            deferred.resolve();
        }

        return deferred.promise;
    };

    /**
     *
     */
    Switch.prototype.setState = function (state) {
        this.state = state;

        this.publishStateChange();
    };

    /**
     *
     */
    Switch.prototype.getState = function () {
        return this.state;
    };

    /**
     *
     */
    Switch.prototype.on = function () {
        return this.state.switch = true;

        this.publishStateChange();
    };

    /**
     *
     */
    Switch.prototype.off = function () {
        return this.state.switch = false;

        this.publishStateChange();
    };

    /**
     *
     */
    Switch.prototype.toggle = function () {
        return this.state.switch = !this.state.switch;

        this.publishStateChange();
    };
}
