// thing-it-device-osram-lightify/examples/configuration.js
module.exports = {
    label: "Suite 1302 Lighting",
    id: "suite1302Lighting",
    autoDiscoveryDeviceTypes: [/*{
     plugin: "osram-lightify/lightifyGateway",
     confirmRegistration: true,
     persistRegistration: false,
     defaultConfiguration: {host: "192.168.192.10", port: 4000},
     options: {}
     }*/],
    devices: [{
        label: "Lightify Gateway",
        id: "lightifyGateway",
        plugin: "osram-lightify/lightifyGateway",
        configuration: {host: "192.168.192.10", port: 4000},
        actors: [{
            id: "lightBulbLounge",
            label: "Light Bulb Lounge",
            type: "lightBulb",
            configuration: {
                mac: '0a:d4:c9:00:00:26:18:84'
            }
        }, {
            id: "led1Kitchen",
            label: "LED 1 Kitchen",
            type: "led",
            configuration: {
                mac: '87:61:01:00:00:26:18:84'
            }
        }, {
            id: "led2Kitchen",
            label: "LED 2 Kitchen",
            type: "led",
            configuration: {
                mac: 'aa:5b:01:00:00:26:18:84'
            }
        }],
        sensors: []
    }],
    groups: [],
    services: [{
        id: "blackout",
        label: "Blackout",
        type: "script",
        content: {
            script: ""
        }
    }],
    eventProcessors: [],
    data: []
};
