// thing-it-device-foscam/camera.js
module.exports = {
    metadata: {
        family: "camera",
        plugin: "camera",
        label: "FOSCAM © Camera",
        tangible: true,
        discoverable: false,
        state: [{
            id: "video",
            label: "Video",
            type: {
                id: "string"
            }
        },{
            id: "videoCGI",
            label: "Video CGI",
            type: {
                id: "string"
            }
        },{
            id: "snapShotImage",
            label: "Snap Shot Image",
            type: {
                id: "string"
            }
        },{
            id: "alias",
            label: "Alias",
            type: {
                id: "string"
            }
        },{
            id: "alarmStatus",
            label: "Alarm Status",
            type: {
                id: "string"
            }
        },{
            id: "temperature",
            label: "Temperature",
            type: {
                id: "decimal"
            }
        },{
            id: "humidity",
            label: "Humidity",
            type: {
                id: "integer"
            }
        }],
        actorTypes: [],
        sensorTypes: [],
        services: [
            {id: "panLeft", label: "Pan Left"},
            {id: "panRight", label: "Pan Right"},
            {id: "tiltUp", label: "Tilt Up"},
            {id: "tiltDown", label: "Tilt Down"},
            {id: "savePosition", label: "Save Position"},
            {id: "restorePosition", label: "Goto Position"}
        ],
        configuration: [{
            id: "host",
            label: "Host",
            type: {id: "string"}
        }, {
            id: "port",
            label: "Port",
            type: {id: "string"}
        }, {
            id: "user",
            label: "User",
            type: {id: "string"}
        }, {
            id: "password",
            label: "Password",
            type: {id: "string"}
        }, {
            id: "updateTime",
            label: "Update Time",
            type: {id: "integer"},
            default: "2000"
/*
        Removing hls and webM live streaming options as they are not needed, but keeping the code that makes it work
        in case we have future need.
        }, {
            id: "hlsLiveStreaming",
            label: "HLS Live Streaming",
            type: {id: "boolean"},
            default: false
        }, {
            id: "webMLiveStreaming",
            label: "webM Live Streaming",
            type: {id: "boolean"},
            default: false
        }, {
            id: "webMPort",
            label: "webM Port",
            type: {id: "integer"},
            default: "3002"
*/
        }]
    },
    create: function () {
        return new Camera();
    },
    discovery: function (options) {
        var discovery = new CameraDiscovery();

        discovery.options = options;

        return discovery;
    }
};

var q = require('q');
var fs = require('fs');
var https = require('https');
var child_process = require('child_process');
var foscamAPI;
var ffmpeg;
var ffserver;
var deviceDataDirectory;
var moveDuration = 200;
var playlistName = "hls-playlist.m3u8";
var ffserverConfName = "ffserver.conf";
var httpReplaceString = "THINGITHTTPPORT";
var deviceReplaceString = "THINGITDEVICE";
var intervals;
var simulationIntervals;


function mkdirSynch(path) {
    var deferred = q.defer();

    try {
        fs.mkdirSync(path);
    } catch (e) {
        if (e.code != 'EEXIST') throw e;
    }

    deferred.resolve();
    return deferred.promise;
}

function CameraDiscovery() {
    var discoveryInterval;
    var vendors;
    var checkedIps;

    CameraDiscovery.prototype.start = function () {
        vendors = {};
        checkedIps = {};

        if (!this.node.isSimulated()) {
            // TODO For now, need to be able to switch for Discovery or inherit from Device
            this.logLevel = "debug";
            this.scanForCameras();
            discoveryInterval = setInterval(this.scanForCameras.bind(this), 30000);
        }
    }

    CameraDiscovery.prototype.stop = function () {
        if (discoveryInterval !== undefined && discoveryInterval) {
            clearInterval(discoveryInterval);
        }
    }

    CameraDiscovery.prototype.scanForCameras = function () {
        this.scanLocalAreaNetworkHosts(function (error, ip, mac) {
            if (error) {
                this.logError(error);
            } else if (ip && !mac) {
                this.logDebug("Ignoring host with unknown MAC address.");
            } else {
                this.getVendorForMac(mac, function (err, vendor) {
                    if (err) {
                        this.logError(ip, err);
                    } else if (vendor) {
                        checkedIps[ip] = true;

                        if (-1 < vendor.indexOf("Shenzhen Reecam")) {
                            this.logInfo("Vendor match found for host "
                                + ip + ": \"" + vendor + "\"");
                            this.testCameraConnection(ip, mac);
                        }
                        else {
                            this.logDebug("Ignoring host " + ip + " with vendor " + vendor + ".");
                        }
                    } else {
                        checkedIps[ip] = true;
                        this.logDebug("No vendor found for IP " + ip + ", trying to connect anyway.");
                        this.testCameraConnection(ip, mac);
                    }
                }.bind(this));
            }
        }.bind(this));
    }

    /**
     * Searches the ARP entries for MAC addresses in the same subnet as the computer the code is running on.
     *
     * @param callback Invokes the callback for each identified MAC address. First parameter is an error,
     *                  second the ip address, third the host.
     */
    CameraDiscovery.prototype.scanLocalAreaNetworkHosts = function (callback) {
        var arp = require('node-arp');
        var network = require('network');

        network.get_active_interface(function (err, obj) {
            if (err) {
                this.logError(err);
                callback(err, false, false);
            } else {
                var myIp = obj.ip_address;
                this.logDebug("My IP address determined as " + myIp);
                var subnet = myIp.substring(0, myIp.lastIndexOf(".") + 1);
                this.logDebug("Scanning for hosts in IP range " + subnet + "1 to " + subnet + "254.");

                for (var i = 1; i < 255; i++) {
                    (function () {
                        var currentIP = subnet + i;

                        if (currentIP != myIp) {
                            if (!checkedIps[currentIP]) { // don't check the same host twice.
                                arp.getMAC(currentIP, function (err, mac) {
                                    if (err) {
                                        this.logError(currentIP, err);
                                    } else {
                                        if (mac !== undefined && mac && ("(incomplete)" != mac) && ("eth0" != mac)) {
                                            this.logDebug("MAC address found for IP " + currentIP + ": " + mac);
                                            callback(false, currentIP, mac);
                                        } else {
                                            checkedIps[currentIP] = false;
                                        }
                                    }
                                }.bind(this));
                            }
                        }
                    }.bind(this))();
                }
            }
        }.bind(this))
    }

    /**
     * Retrieves the vendor name for a MAC address.
     * @param mac       The MAC address to look up.
     * @param callback  Invoked with return values with first parameter an error, second the vendor name.
     */
    CameraDiscovery.prototype.getVendorForMac = function (mac, callback) {
        this.logDebug("Looking up vendor for MAC address " + mac + ".");
        var cashedVendor = vendors[mac];

        if (cashedVendor || ("" == cashedVendor)) {
            this.logDebug("Retrieved vendor '" + cashedVendor + "' from vendor cache.");

            if ("" !== cashedVendor) {
                callback(false, cashedVendor);
            } else {
                callback(false, false);
            }
        } else {
            var data = "";

            var options = {
                hostname: 'www.macvendorlookup.com',
                port: 443,
                path: '/api/v2/' + mac,
                method: 'GET',
                rejectUnauthorized: false
            };

            var req = https.request(options, function (res) {
                res.on('data', function (chunk) {
                    data += chunk;
                }.bind(this));

                res.on('end', function () {
                    var vendor;

                    if ("" !== data) {
                        var vendorLookup = JSON.parse(data);

                        if ((vendorLookup !== undefined && vendorLookup) &&
                            (vendorLookup[0] !== undefined && vendorLookup[0]) &&
                            (vendorLookup[0].company !== undefined && vendorLookup[0].company)) {
                            vendor = vendorLookup[0].company;
                            this.logDebug("Vendor for MAC address " + mac + " identified as " + vendor + ".");
                            vendors[mac] = vendor;
                            callback(false, vendor);
                        }
                    } else {
                        this.logDebug("Vendor for MAC address " + mac + " not identified.");
                        vendors[mac] = "";
                        callback(false, false);
                    }
                }.bind(this));
            }.bind(this));

            req.end();

            req.on('error', function (e) {
                this.logError("MAC based vendor lookup failed for MAC address " + mac + ".", e);
                callback(e, false);
            }.bind(this));
        }

    }

    CameraDiscovery.prototype.testCameraConnection = function (host, mac) {
        this.logDebug("Testing connection to host " + host + ".");
        var discoveryFoscam = require('foscam');

        discoveryFoscam.setup({
            host: host,
            port: 80,
            user: "admin",
            pass: ""
        }, function (status) {
            if (typeof status === undefined || !status || (status.undefined && status.undefined === "undefined")) {
                this.logError("Can't connect to suspected FOSCAM camera host at " + host + " with default credentials.");

                discoveryFoscam.setup({
                    host: host,
                    port: 80,
                    user: "thingit",
                    pass: "[th1ng-1t]"
                }, function (status) {
                    if (typeof status === undefined || !status || (status.undefined && status.undefined === "undefined")) {
                        this.logError("Can't connect to FOSCAM camera at " + host + " with [thing-it] credentials." +
                            " Host skipped. ");
                    }
                    else if (status.alias !== undefined && status.alias) {
                        this.logInfo("Confirmed host " + host + " as FOSCAM camera with alias " + status.alias + ".");
                        this.advertiseCamera(status.alias, host, 80, "thingit", "[th1ng-1t]", mac);
                    }
                    else {
                        this.logError("Retrieved unexpected result from suspected FOSCAM camera at " + host +
                            ". Host skipped.");
                        this.logDebug(status);
                    }
                }.bind(this));
            }
            else if (status.alias !== undefined && status.alias) {
                this.logInfo("Confirmed host " + host + " as FOSCAM camera with alias " + status.alias + ".");
                this.advertiseCamera(status.alias, host, 80, "admin", "", mac);
            }
            else {
                this.logError("Retrieved unexpected result from suspected FOSCAM camera at " + host +
                    ". Host skipped.");
                this.logDebug(status);
            }
        }.bind(this));

    }

    CameraDiscovery.prototype.advertiseCamera = function (alias, host, port, user, pass, mac) {
        var foscamCamera = new Camera();
        foscamCamera.id = "FoscamCamera" + alias.replace(/\W/g, '');
        foscamCamera.label = alias;
        foscamCamera.uuid = mac;
        foscamCamera.configuration = {};
        foscamCamera.configuration.host = host;
        foscamCamera.configuration.port = port;
        foscamCamera.configuration.user = user;
        foscamCamera.configuration.password = pass;
        foscamCamera.configuration.hlsLiveStreaming = false;
        foscamCamera.configuration.webMLiveStreaming = false;
        this.logDebug("FOSCAM camera", foscamCamera);
        this.logInfo("Advertising camera \"" + alias + "\" with UUID " + mac + ".");
        this.advertiseDevice(foscamCamera);
    }
}

/**
 *
 */
function Camera() {
    /**
     *
     */
    Camera.prototype.start = function () {
        var promise;
        this.state = {};
        this.intervals = [];
        this.simulationIntervals = [];
        this.logDebug(__dirname);

        if (this.configuration.hlsLiveStreaming === undefined) {
            this.configuration.hlsLiveStreaming = false;
        }

        if (this.configuration.webMLiveStreaming === undefined) {
            this.configuration.webMLiveStreaming = false;
        }

        if (this.configuration.webMPort === undefined || !this.configuration.webMPort) {
            this.configuration.webMPort = 3002;
        }

        if (this.configuration.updateTime === undefined || !this.configuration.updateTime) {
            this.configuration.updateTime = 5000;
        }

        // Create directory for snapshot images and streaming buffer
        if (this.node.options.dataDirectory !== undefined && this.node.options.dataDirectory) {
            deviceDataDirectory = this.node.options.dataDirectory + "/devices/" + this.id;
            mkdirSynch(this.node.options.dataDirectory);
            mkdirSynch(this.node.options.dataDirectory + "/devices");
            mkdirSynch(deviceDataDirectory);

            this.imageIndex = 0;
            this.rotateIndex = 0;

            this.configuration.port = this.configuration.port || 80;
            this.configuration.user = this.configuration.user || "admin";
            this.configuration.password = this.configuration.password || "";

            if (this.isSimulated()) {
                promise = q.fcall(function () {
                    return "Started in simulated mode."
                });
            } else {
                this.logDebug("Starting FOSCAM camera with configuration ", this.configuration);

                if (!foscamAPI) {
                    foscamAPI = require('./lib/foscamAPI').create();
                }

                promise = foscamAPI.connect(this.configuration.host, this.configuration.port, this.configuration.user, this.configuration.password)
                    .then(function (result) {
                        this.foscamStatus = result;
                        this.state.alias = result.alias;
                        this.state.alarmStatus = result.alarm_status;
                        this.state.temperature = parseFloat(result.temperature);
                        this.state.humidity = parseInt(result.humidity);
                        this.logInfo("Connected to FOSCAM camera with alias " + this.foscamStatus.alias +
                            " at IP address " + this.configuration.host + ".");
                        this.state.videoCGI = foscamAPI.getVideoCGIUrl();
                        this.logDebug("Video CGI: " + this.state.videoCGI);
                        this.flipImage();
                        this.intervals.push(setInterval(function () {
                                this.flipImage()
                            }.bind(this),
                            this.configuration.updateTime));

                        // this.startStreaming();
                    }.bind(this), function (error) {
                        this.logError(error);
                    }.bind(this));
            }
        } else {
            this.logError("Data directory required for video streaming files not defined, cannot start FOSCAM device.");
            promise = q.fcall(function () {
                return "Data directory required for video streaming files not defined, cannot start FOSCAM device.";
            });
        }


        return promise;
    };

    /**
     *
     *
     */
    Camera.prototype.stop = function () {
        this.logDebug("FOSCAM stop called");
        clearInterval(this.flipImageInterval);

        for (var interval in this.intervals) {
            clearInterval(interval);
        }

        for (var interval in this.simulationIntervals) {
            clearInterval(interval);
        }

        if (ffmpeg) {
            ffmpeg.kill();
        }
        if (ffserver) {
            ffserver.kill();
        }
    };

    Camera.prototype.startStreaming = function () {
        var deferred = q.defer();

        if (this.configuration.webMLiveStreaming || this.configuration.hlsLiveStreaming) {

            var ffmpegVariables = [
                "-i", "http://" + this.configuration.user + ":" + this.configuration.password + "@" +
                this.configuration.host + ":" + this.configuration.port + "/videostream.asf"
            ];

            if (this.configuration.hlsLiveStreaming) {
                this.state.playlist = "/data/devices/" + this.id + "/" + playlistName;

                fs.createReadStream(__dirname + "/data/" + playlistName).pipe(
                    fs.createWriteStream(deviceDataDirectory + "/" + playlistName));

                ffmpegVariables = ffmpegVariables.concat([
                    "-c:a", "aac", "-strict", "experimental", "-ac", "2",
                    "-b:a", "96k", "-ar", "44100", "-c:v", "libx264",
                    "-pix_fmt", "yuv420p", "-preset", "ultrafast", "-profile:v",
                    "baseline", "-level", "1.3", "-maxrate", "192k",
                    "-bufsize", "1M", "-crf", "18", "-r", "24", "-g", "72",
                    "-f", "hls", "-hls_time", "1", "-hls_list_size", "2",
                    "-hls_wrap", "4", "-s", "320x240", deviceDataDirectory + "/320x240.m3u8",
                    "-c:a", "aac", "-strict", "experimental", "-ac", "2",
                    "-b:a", "96k", "-ar", "44100", "-c:v", "libx264",
                    "-pix_fmt", "yuv420p", "-preset", "ultrafast", "-profile:v",
                    "baseline", "-level", "3.1", "-maxrate", "1M",
                    "-bufsize", "3M", "-crf", "18", "-r", "24", "-g", "72",
                    "-f", "hls", "-hls_time", "1", "-hls_list_size", "2",
                    "-hls_wrap", "4", "-s", "640x480", deviceDataDirectory + "/640x480.m3u8"
                ]);
            }

            if (this.configuration.webMLiveStreaming) {
                fs.readFile(__dirname + "/data/" + ffserverConfName, 'utf8', function (err, data) {
                    var deferred = q.defer();

                    if (err) {
                        deferred.resolve();
                        this.logError(err);
                    } else {
                        var result = data.replace(/THINGITHTTPPORT/g, this.configuration.webMPort);
                        result = result.replace(/THINGITDEVICE/g, this.id);

                        fs.writeFile(deviceDataDirectory + "/" + ffserverConfName, result, 'utf8', function (err) {
                            var deferred = q.defer();

                            if (err) {
                                deferred.resolve();
                                this.logError("Error creating FFserver configuration.", err);
                            }
                            else {
                                this.logDebug("Successfully written FFserver configuration.");

                                ffserver = child_process.spawn("ffserver", [
                                    "-f", deviceDataDirectory + "/" + ffserverConfName
                                ], {detached: false});

                                ffserver.stdout.on("data", function (data) {
                                    this.logDebug("FFserver received data.", data);
                                }.bind(this));

                                ffserver.stderr.on("data", function (data) {
                                    this.logDebug("FFserver Standard Error Out -> " + data);
                                }.bind(this));

                                ffserver.on("exit", function (code) {
                                    this.logError("FFserver terminated with code " + code);
                                }.bind(this));

                                ffserver.on("error", function (e) {
                                    this.logError("FFserver system error: " + e);
                                }.bind(this));
                                deferred.resolve();
                            }
                        }.bind(this));

                        deferred.resolve();
                    }

                    return deferred.promise;
                }.bind(this));

                var webMTarget = "http://localhost:" + this.configuration.webMPort + "/" + this.id + ".ffm";
                this.logDebug(webMTarget);
                ffmpegVariables = ffmpegVariables.concat(webMTarget);
            }

            if (ffmpeg) {
                this.logInfo("Stopping existing FFmpeg process.");
                ffmpeg.kill();
            }

            ffmpeg = child_process.spawn("ffmpeg", ffmpegVariables, {detached: false});

            ffmpeg.stdout.on("data", function (data) {
                this.logDebug("FFmpeg received data.");
                this.writeData(data);
            }.bind(this));

            ffmpeg.stderr.on("data", function (data) {
                this.logDebug("FFmpeg Standard Error Out -> " + data);
            }.bind(this));

            ffmpeg.on("exit", function (code) {
                this.logError("FFmpeg terminated with code " + code);
            }.bind(this));

            ffmpeg.on("error", function (e) {
                this.logError("FFmpeg system error: " + e);
            }.bind(this));

            deferred.resolve();
        } else {
            deferred.resolve();
            this.logDebug("Streaming not requested - skipping FFmpeg and FFserver setup.");
        }

        return deferred.promise;
    }

    /**
     *
     */
    Camera.prototype.flipImage = function () {
        var deferred = q.defer();
        this.imageIndex = (this.imageIndex + 1) % 3;
        this.rotateIndex = (this.rotateIndex + 1) % 2;

        if (!this.isSimulated()) {
            this.logDebug("Retrieving next snapshot.");
            this.state.snapshotImage = "/data/devices/" + this.id + "/snapshotImage" + this.rotateIndex + ".jpg";
            var fileName = deviceDataDirectory + '/snapshotImage' + this.rotateIndex + '.png';

            foscamAPI.getSnapShot().then(function (binary) {
                fs.writeFile(fileName, binary, "binary", function (err) {
                    if (err) {
                        deferred.reject(err);
                    } else {
                        deferred.resolve();
                        this.logDebug("Snapshot retrieved.");
                        this.publishStateChange();
                    }
                }.bind(this));
            }.bind(this));
        } else {
            this.logDebug("Simulation of flipImage.");
            fs.createReadStream(__dirname + "/data/snapshotImage" + this.imageIndex + ".png").pipe(
                fs.createWriteStream(deviceDataDirectory + "/snapshotImage" + this.rotateIndex + ".png"));
            deferred.resolve();
        }

        return deferred.promise;
    }

    /**
     *
     */
    Camera.prototype.setState = function (state) {
        this.state = state;
    };

    /**
     *
     */
    Camera.prototype.getState = function () {
        return this.state;
    };

    /**
     *
     */
    Camera.prototype.video = function (req, res) {
        if (this.isSimulated()) {
            this.logDebug("Piping file");
            this.pipeFile(req, res, __dirname + "/data/spaceship.m4v", "video/mp4");
        } else {
            this.logDebug("Piping playlist file for HLS.");
            this.pipeFile(req, res, deviceDataDirectory + "/hls-playlist.m3u4", "application/x-mpegURL");
        }
    };

    Camera.prototype.moveCamera = function (direction) {
        return this.sendControl(direction, true);
    };

    Camera.prototype.sendControl = function (control, oneStep) {
        var deferred = q.defer();

        if (this.isSimulated()) {
            this.flipImage();
            deferred.resolve();
        } else {
            this.logDebug('Sending camera control command (' + control + ').');
            foscamAPI.ptzControl(control, oneStep)
                .then(function () {
                    this.flipImage();
                }.bind(this))
                .fail(function (error) {
                    this.logError(error);
                }.bind(this))
                .done(function () {
                    this.logDebug('Command (' + control + ') sent.');
                    deferred.resolve();
                }.bind(this));
        }

        return deferred.promise;
    };

    /**
     *
     */
    Camera.prototype.panLeft = function () {
        return this.moveCamera(foscamAPI.controls.left);
    };

    /**
     *
     */
    Camera.prototype.panRight = function () {
        return this.moveCamera(foscamAPI.controls.right);
    };

    /**
     *
     */
    Camera.prototype.tiltUp = function () {
        return this.moveCamera(foscamAPI.controls.up);
    };

    /**
     *
     */
    Camera.prototype.tiltDown = function () {
        return this.moveCamera(foscamAPI.controls.down);
    };

    /**
     *
     */
    Camera.prototype.savePosition = function () {
        return this.sendControl(foscamAPI.controls.setPreset1);
    };

    /**
     *
     */
    Camera.prototype.restorePosition = function () {
        return this.sendControl(foscamAPI.controls.goToPreset1);
    };

    /**
     *
     */
    Camera.prototype.getResourceUrl = function (stateVariable) {
        if (this.isSimulated()) {
            return __dirname + "/data/spaceship.m4v"
        } else {
            return 'http://' + this.configuration.host + ':' + this.configuration.port + '/videostream.cgi?user=' + this.configuration.user +
                '&pwd=' + this.configuration.password;
        }
    };
}
