// thing-it-device-rfid-sensor/rfidSensor.js
module.exports = {
    metadata: {
        family: 'rfidSensor',
        plugin: 'rfidSensor',
        label: 'Generic RFID Sensor',
        manufacturer: 'Generic',
        discoverable: false,
        tangible: true,
        additionalSoftware: [],
        actorTypes: [],
        sensorTypes: [],
        state: [{
            id: "lastRfidCode",
            label: "Last RFID Code",
            type: {
                id: "string"
            }
        }],
        services: [],
        configuration: []
    },
    create: function () {
        return new RfidSensor();
    },
    discovery: function () {
        return new RfidSensorDiscovery();
    }
};

var q = require('q');
var _ = require('lodash');

function RfidSensorDiscovery() {
    RfidSensorDiscovery.prototype.start = function () {

        if (!this.node.isSimulated()) {
            // TODO For now, need to be able to switch for Discovery or inherit from Device
            this.logLevel = "debug";
        }
    };

    RfidSensorDiscovery.prototype.stop = function () {
        if (discoveryInterval !== undefined && discoveryInterval) {
            clearInterval(discoveryInterval);
        }
    };
}

/**
 *
 * @constructor
 */
function RfidSensor() {
    RfidSensor.prototype.start = function () {
        var deferred = q.defer();

        this.state = {};

        if (this.isSimulated()) {
            this.statusChangeInterval = setInterval(function () {
                var tagIds = ["XD345677YH8",
                    "JJKL8885674",
                    "00765FFG3X2"];
                this.publishEvent('tagIdentified', {details: this.state.lastRfidCode = tagIds[new Date().getTime() % tagIds.length]});
                this.publishStateChange();
            }.bind(this), 20000);

            deferred.resolve();
        } else {

            deferred.resolve();
        }

        return deferred.promise;
    };

    /**
     *
     */
    RfidSensor.prototype.stop = function () {
        var deferred = q.defer();

        if (this.isSimulated()) {
            if (this.statusChangeInterval) {
                clearInterval(this.statusChangeInterval);
            }
        } else {
        }

        deferred.resolve();

        return deferred.promise;
    };

    /**
     *
     */
    RfidSensor.prototype.getState = function () {
        return this.state;
    };

    /**
     *
     */
    RfidSensor.prototype.setState = function () {
    };
}
